Code/GraphMol/FileParsers/MaeMolSupplier.cpp

```diff
@@ -10,243 +10,424 @@
+    if (m_resnum && m_resnum->isDefined(atom_num)) {
+      rd_info->setResidueNumber(m_resnum->at(atom_num));
+    }
 
-};
+    if (m_occupancy && m_occupancy->isDefined(atom_num)) {
+      rd_info->setOccupancy(m_occupancy->at(atom_num));
+    }
 
-PDBInfo::PDBInfo(const shared_ptr<const IndexedBlock>& atom_data)
-{
-    try {
-    m_atom_name = atom_data->getStringProperty("s_m_pdb_atom_name");
-    } catch (std::out_of_range &) { }
+    if (m_tempfac && m_tempfac->isDefined(atom_num)) {
+      rd_info->setTempFactor(m_tempfac->at(atom_num));
+    }
+  }
 
-    try {
-    m_residue_name = atom_data->getStringProperty("s_m_pdb_residue_name");
-    } catch (std::out_of_range &) { }
+ private:
+  std::shared_ptr<mae::IndexedStringProperty> m_atom_name;
+  std::shared_ptr<mae::IndexedStringProperty> m_residue_name;
+  std::shared_ptr<mae::IndexedStringProperty> m_chain_id;
+  std::shared_ptr<mae::IndexedStringProperty> m_insertion_code;
 
-    try {
-    m_chain_id = atom_data->getStringProperty("s_m_chain_name");
-    } catch (std::out_of_range &) { }
+  std::shared_ptr<mae::IndexedIntProperty> m_resnum;
 
-    try {
-    m_insertion_code = atom_data->getStringProperty("s_m_insertion_code");
-    } catch (std::out_of_range &) { }
+  std::shared_ptr<mae::IndexedRealProperty> m_occupancy;
+  std::shared_ptr<mae::IndexedRealProperty> m_tempfac;
+};
 
-    try {
-    m_resnum = atom_data->getIntProperty("i_m_residue_number");
-    } catch (std::out_of_range &) { }
+bool streamIsGoodOrExhausted(std::istream *stream) {
+  PRECONDITION(stream, "bad stream");
+  return stream->good() || (stream->eof() && stream->fail() && !stream->bad());
+}
 
-    try {
-    m_occupancy = atom_data->getRealProperty("r_m_pdb_occupancy");
-    } catch (std::out_of_range &) { }
+void parseChiralityLabel(RWMol &mol, const std::string &stereo_prop) {
+  boost::char_separator<char> sep{"_"};
+  boost::tokenizer<boost::char_separator<char>> tokenizer{stereo_prop, sep};
+
+  auto tItr = tokenizer.begin();
+
+  const int chiral_idx = FileParserUtils::toInt(*tItr) - 1;
+  Atom *chiral_atom = mol.getAtomWithIdx(chiral_idx);
+  CHECK_INVARIANT(chiral_atom != nullptr, "bad prop value");
+
+  unsigned nSwaps = 2;
+  switch (*((++tItr)->rbegin())) {
+    case 'R':  // R, ANR
+      nSwaps = 0;
+      break;
+    case 'S':  // S, ANS
+      nSwaps = 1;
+      break;
+    default:
+      break;
+  }
+  CHECK_INVARIANT(nSwaps < 2, "bad prop value");
 
-    try {
-    m_tempfac = atom_data->getRealProperty("r_m_pdb_tfactor");
-    } catch (std::out_of_range &) { }
+  INT_LIST bond_indexes;
+  for (++tItr; tItr != tokenizer.end(); ++tItr) {
+    const int nbr_idx = FileParserUtils::toInt(*tItr) - 1;
+    const Bond *bnd = mol.getBondBetweenAtoms(chiral_idx, nbr_idx);
+    bond_indexes.push_back(bnd->getIdx());
+  }
+  CHECK_INVARIANT(bond_indexes.size() == chiral_atom->getDegree(),
+                  "bad prop value");
+
+  nSwaps += chiral_atom->getPerturbationOrder(bond_indexes);
```

Comment:
What about hydrogens? If this is a chirality, any hydrogen would be at the end. But if this were a molecule with explicit hydrogens and atom-numbering-only absolute configuration, the hydrogen wouldn't necessarily be last in the mae property.

Reply:
At the point of doing this, all atoms are still explicit: in the `next()` method, first, atoms are parsed, then bonds, then mol properties, and then, Hydrogen nodes are removed, if we opted to do so.